sled-agent/src/instance.rs

```diff
@@ -186,9 +198,117 @@ struct PropolisSetup {
     running_zone: RunningZone,
 }
 
-struct InstanceInner {
+// Requests that can be made of instances
+enum InstanceRequest {
+    RequestZoneBundle {
+        tx: oneshot::Sender<Result<ZoneBundleMetadata, BundleError>>,
+    },
+    CurrentState {
+        tx: oneshot::Sender<SledInstanceState>,
+    },
+    PutState {
+        state: crate::params::InstanceStateRequested,
+        tx: oneshot::Sender<Result<InstancePutStateResponse, ManagerError>>,
+    },
+    PutMigrationIds {
+        old_runtime: InstanceRuntimeState,
+        migration_ids: Option<InstanceMigrationSourceParams>,
+        tx: oneshot::Sender<Result<SledInstanceState, ManagerError>>,
+    },
+    Terminate {
+        tx: oneshot::Sender<Result<InstanceUnregisterResponse, ManagerError>>,
+    },
+    IssueSnapshotRequest {
+        disk_id: Uuid,
+        snapshot_id: Uuid,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+    AddExternalIp {
+        ip: InstanceExternalIpBody,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+    DeleteExternalIp {
+        ip: InstanceExternalIpBody,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+}
+
+// A small task which tracks the state of the instance, by constantly querying
+// the state of Propolis for updates.
+//
+// This task communicates with the "InstanceRunner" task to report status.
+struct InstanceMonitorRunner {
+    client: Arc<PropolisClient>,
+    tx_monitor: mpsc::Sender<InstanceMonitorRequest>,
+}
+
+impl InstanceMonitorRunner {
+    async fn run(self) -> Result<(), anyhow::Error> {
+        let mut gen = 0;
+        loop {
+            // State monitoring always returns the most recent state/gen pair
+            // known to Propolis.
+            let response = self
+                .client
+                .instance_state_monitor()
+                .body(propolis_client::types::InstanceStateMonitorRequest {
+                    gen,
+                })
+                .send()
+                .await
+                .map_err(|e| anyhow!(e))?
+                .into_inner();
+            let observed_gen = response.gen;
+
+            // Now that we have the response from Propolis' HTTP server, we
+            // forward that to the InstanceRunner.
+            //
+            // It will decide the new state, provide that info to Nexus,
+            // and possibly identify if we should terminate.
+            let (tx, rx) = oneshot::channel();
+            self.tx_monitor
+                .send(InstanceMonitorRequest::Update { state: response, tx })
+                .await
+                .map_err(|e| anyhow!(e))?;
+
+            if let Reaction::Terminate = rx.await.map_err(|e| anyhow!(e))? {
```

Comment:
and i think this can probably also just be 
```suggestion
            if let Reaction::Terminate = rx.await? {
```

Reply:
Updated